src/platform_linux/laio.c

```diff
@@ -378,34 +518,58 @@ laio_cleanup(io_handle *ioh, uint64 count)
    uint64          i;
    int             status;
 
+   threadid tid = platform_get_tid();
+
    io = (laio_handle *)ioh;
 
    // Check for completion of up to 'count' events, one event at a time.
    // Or, check for all outstanding events (count == 0)
    for (i = 0; ((count == 0) || (i < count)); i++) {
-      status = io_getevents(io->ctx, 0, 1, &event, NULL);
+      status = io_getevents(io->ctx[tid], 0, 1, &event, NULL);
       if (status < 0) {
-         platform_error_log("io_getevents[%lu], count=%lu, "
-                            "failed with errorno=%d: %s\n",
-                            i,
-                            count,
-                            -status,
-                            strerror(-status));
+         platform_error_log(
+            "%s(): OS-pid=%d, tid=%lu, io_getevents[%lu], count=%lu, "
+            "failed with errorno=%d: %s\n",
+            __func__,
+            getpid(),
+            tid,
+            i,
+            count,
+            -status,
+            strerror(-status));
          i--;
-         continue;
+
+         // We got a hard-error probably because some code-flow messed-up
+         // using the per-thread IO-context. No point in trying again and
+         // again if we are checking for completion of all outstanding
+         // events.
+         if (count == 0) {
+            break;
+         } else {
+            // Retry a few times for a finite # of pending events
+            continue;
+         }
       }
       // No event has completed, so we are done. Exit.
       if (status == 0) {
          break;
       }
       // Invoke the callback for the one event that completed.
-      laio_callback(io->ctx, event.obj, event.res, 0);
+      laio_callback(io->ctx[tid], event.obj, event.res, 0);
    }
 }
 
 /*
  * laio_cleanup_all() - Handle completion of outstanding IO requests,
  * for all async requests in the queue.
+ *
+ * RESOLVE: Note this interface is potentially problematic when deployed
+ * using a process model. The expectation is that only the parent process
+ * will invoke this to do a final cleanup. But if there are any pending
+ * aios left behind by a process that exited before deregistering itself
+ * cleanly, we may find that AIO-request busy below, and will try to call
+ * the cleanup function. As the IO-context is no longer correct, we will
+ * get a hard-error from io_getevents() call.
```

Comment:
ATTN: Let's discuss this point ... I had some misgivings in this area, so put down this comment.

We should talk ... to see if there are any latent problems lurking here ... ???